File: practice/binary_tree/binary_tree_preorder_traversal.py
# binary_tree_preorder_traversal.py --- 
# 
# Filename: binary_tree_preorder_traversal.py
# Description: 
# 
# Author:    Yu Lu
# Email:     yulu@utexas.edu
# Github:    https://github.com/SuperYuLu 
# 
# Created: Sun Oct 21 16:25:07 2018 (-0500)
# Version: 
# Last-Updated: Sun Oct 21 16:42:04 2018 (-0500)
#           By: yulu
#     Update #: 6
# 

class Solution:
    '''
    Given a binary tree, return the preorder traversal of its nodes' values.

    Definition for a binary tree node.
    class TreeNode:
        def __init__(self, x):
            self.val = x
            self.left = None
            self.right = None

    '''
    
    #Recursive solution
    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        if root is None:     
            return []
        else:
            return [root.val, *self.preorderTraversal(root.left), *self.preorderTraversal(root.right)]
     
    #Iterative solution
    def preorderTraversal(self, root):
        res = []
        stack = []
        while root:
            res.append(root.val)
            if root.right:
                stack.append(root.right)
            if root.left:
                stack.append(root.left)
            try:
                root = stack.pop()
            except IndexError:
                break
            
        return res
         

File: practice/binary_tree/binary_tree_preorder_traversal.cc
// preorderTraversal.cc --- 
// 
// Filename: preorderTraversal.cc
// Description: 
// 
// Author:    Yu Lu
// Email:     yulu@utexas.edu
// Github:    https://github.com/SuperYuLu 
// 
// Created: Thu Nov 22 14:38:19 2018 (-0600)
// Version: 
// Last-Updated: Thu Nov 22 14:38:46 2018 (-0600)
//           By: yulu
//     Update #: 1
// 

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution {
public:
  vector<int> preorderTraversal(TreeNode* root) {
    vector<int> res;
    if(root) preTraversal(root, res);
    return res;
  }   
  void preTraversal(TreeNode* root, vector<int>& v){
    v.push_back(root -> val);
    if(root -> left) preTraversal(root -> left, v);
    if(root -> right) preTraversal(root -> right, v);
  }
};
